### PR TITLE
AIP-44 Make standalone dag file processor works in DB isolation mode

### PR DESCRIPTION
**Pull request by potiuk**
There were a few missing DB operations in DAGFileProcessor that prevented it to run in DB isolation mode. Those have been refactored and exposed as internal API calls.

A bug was fixed in scheduler_job_runner that caused using of next_event before it has been declared (which occured when standalone dag processor is used and db isolation mode.

The DB retry will now correctly use logger when it is used as decorator on class method.

The "main" code that removes DB connection from configuration (mostly in case of Breeze) when untrusted components are used has been improved to handle the case where DAGFile Processor forks parsing subprocesses.

Tmux configuration got improved so that both non-isolation and isolation mode distribute panels better.

Simplified InternalApiConfig - "main" directly sets db/internal use in db_isolation mode depending on the component.

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by potiuk** [comment]
There are some improvements here in InternalApiConfig implementation which is much simpler now @jscheffl and @dstandish. 

So now choosing whether to use Internal API (trusted components only) or DB (if DB configuration is available in DB isolation mode  - like it is in Breeze environment) is purely done in main() + we seet the DB configuration to "none://" - which is handled in case there are forks that are re-inititalizing ORM.

**Comment by potiuk** [comment]
There is a chance It's going to be green @jscheffl @vincbeck @dstandish -> I have to do a few more "manual" tests but I think that one is a complete "Internal_api" implementation for standalone dag file processor.

It can be easily tested with breeze:

`breeze start-airflow --load-example-dags --load-default-connections --executor CeleryExecutor --standalone-dag-processor --database-isolation`

Also without isolation:

`breeze start-airflow --load-example-dags --load-default-connections --executor CeleryExecutor --standalone-dag-processor`

When you run airflow in isolation mode, it should nicely log what "mode" each component works in - in isolation mode, scheduler and webserver (and generally all regular airflow commands like `airflow db` etc. switch from isolation mode to "regular DB" mode. This is because in breeze we have both "conn_sql" and "isolation_mode" set and thanks to that we can "mix" isolation/no-isolation mode.

**Comment by potiuk** [comment]
Should be merged after #40929 -> found it during testing standalone dag processor case.

File: .github/workflows/ci.yml
Comment (vincbeck) [inline .github/workflows/ci.yml]: Which means we are going to run tests in DB isolation mode only? Should we run them in both mode?

File: airflow/api_internal/internal_api_call.py
Comment (vincbeck) [inline airflow/api_internal/internal_api_call.py]:
```suggestion
```

File: airflow/api_internal/internal_api_call.py
Comment (vincbeck) [inline airflow/api_internal/internal_api_call.py]:
```suggestion
        url_conf = urlparse(internal_api_url)
```

File: airflow/api_internal/internal_api_call.py
Comment (vincbeck) [inline airflow/api_internal/internal_api_call.py:77]: There is an inconsistency between this check and the one on line 72

File: airflow/api_internal/internal_api_call.py
Comment (vincbeck) [inline airflow/api_internal/internal_api_call.py:78]: `internal_api_endpoint` is already defined

File: airflow/cli/commands/internal_api_command.py
Comment (vincbeck) [inline airflow/cli/commands/internal_api_command.py:226]: Should we fail this command if `database_access_isolation` is `False`? The user is trying to run the internal API component but `database_access_isolation` is `False`

File: .github/workflows/ci.yml
Comment (potiuk) [reply .github/workflows/ci.yml]: Ah. Yeah. I think we can remove it now - those are basic tests (docs/api generation etc. - they do not run the "unit tests"

File: airflow/api_internal/internal_api_call.py
Comment (potiuk) [reply airflow/api_internal/internal_api_call.py:77]: Yeah. Bad merge.

File: airflow/cli/commands/internal_api_command.py
Comment (potiuk) [reply airflow/cli/commands/internal_api_command.py:226]: Yes. Why not.  It would continue to work, byt it's more consistent this way.

File: airflow/dag_processing/processor.py
Comment (vincbeck) [inline airflow/dag_processing/processor.py]: I am curious. You create a session here but you seem not using it (besides for the listener)?

Comment (potiuk) [reply airflow/dag_processing/processor.py]: Yeah. I have to take a closer look at that one - I think I made wrong assumption here about where the queries were counted and I even commented it out - but I see now it's quite a bit differrent counting

**Review by jscheffl — APPROVED** [review]
Except the comments from @vincbeck I have no other questions/objections. Assume manual tests not needed as pipeline is checking all changes to make it green.

**Comment by potiuk** [comment]
> Except the comments from @vincbeck I have no other questions/objections. Assume manual tests not needed as pipeline is checking all changes to make it green.

Still working on best way of solving it.

File: airflow/dag_processing/processor.py
Comment (potiuk) [reply airflow/dag_processing/processor.py]: All right @jscheffl and @vincbeck . I think I have it right - I will do some more manual testing and see if it works in various combinations. Looking for your comments.

But the things I found that the counter was LIKELY not doing what it was supposed to do when added in https://github.com/apache/airflow/pull/40323 

@MaksYermak @kosteev @VladaZakharova -> this is probably last "serious" change to make AIP-44 work (still special tests and docs will be needed but this can be added after that). Let me know if I understood it properly.

When I looked at the counter from #40323 the event was added to the session just before creating dagbag from the file - apparently with the intention to cound the DB queries done inside the dagbag parsing. But as far as I uniderstend (please correct me if I am wrong) ... It would not work because the session the event has been registered for is not passed down/used in the dagbag parsing - any DB access there would create a new session rather than use the session created in "process_file" - so it would not count any queries there. But it was indeed counting potentially queries run during callback, updating warning etc.

I **THINK** that was not the intention of that change - the intention was to count queries resuting from parsing the DAG (but it will not). I am not sure if we can easily fix it (but it does not matter in DB isolation mode at least because no DB queries will be allowed when parsing the DAG anyway). 

In this change I attempted to retain that behaviour. Due to the fact that we cannot pass parsed dag to Internal-api component, the whole "process_file" is now effectively split into more transactons - 4 instead of 2 as it was originally (but I think that is ok, transactional integrity is not important in those cases - and I collect the count of queries and add them up from all 4 calls that might turn into "internal_api_call" - so DB queries will be counted in remote end and passed back as return value of those calls. So the metrics should remain the same as it was.

I introduced a context manager for that - that is nicer than the original method decorated. Speaking of which I also found one potential memory leak. I think when you register for the events, you should always remove the event function because otherwise it might leak memory (see the docstring). It would not matter much in this case - because parsing file is always done in a fork, but I did it in the way that context manager cleans up and removes such registered event - in "database isolation" case it would matter because there memory leaks could stay in the gunicorn threads handling the calls. 

Let me know if the lines of thought above make sense please :)

**Review by vincbeck — APPROVED** [review]
LGTM!

File: airflow/dag_processing/processor.py
Comment (vincbeck) [reply airflow/dag_processing/processor.py]: Makes sense to me. Thanks for sharing the details. I like the context manager solution

**Comment by potiuk** [comment]
Allright @jscheffl @dstandish @vincbeck  (and @mhenc) - it looks like we have it. 

Once i have some one more pair of eyes I think this one concludes the main part (and I think we made it before 2.10 cut-off @ephraimbuddy ?). 

What nicely work there (at least manually checks):

a) airflow in breeze works out-of-the-box with any combination of:
  * `--db-isolation-mode`
  * `--standalone-dag-processor` 
  * Local and Celery Executor (Local Executor mostly because I wanted to test if it works when run K8S Executor that intenrnally uses LocalExecutor

b) Authentication works nicely using separate internal_api_secret_key.

c) Performance does not look bad actually at least on local machine - but I am sure it can be optimized further. 

d) we do not use https:// - but I think this should be described in the documentation that SSL terminatinb proxies shoudl be used in front of internal-api to add SSL (usual practice). We might want to give a big WARNING if someone uses HTTP:// in the client

What we will need more is:

a) special tests for isolatin mode
b) documentation  (stressing experimental status) 

But those can be even added after the cut-off for 2.10 branch and cherry-picked potentially.

I think particularly the Dag File Processor was an interesting one in lights of AIP-72 - I think it can be much more easily mapped almost 1-1 to whatever AIP-72 brings. And nice thing about it that we can tell adventorous users to use the internal API and report any issues they encounter with lack of DB access on the side of worker and parser long before AIP-72 will be out)

**Comment by potiuk** [comment]
ccL @ashb  -> might be useful for AIP-72 ^^

File: airflow/dag_processing/processor.py
Comment (potiuk) [reply airflow/dag_processing/processor.py]: @MaksYermak @kosteev @VladaZakharova -> before I merge this one, please let me know if I correctly assessed the event behaviour. Was the intention to count the queries made by the DAG parsing as well ? Because if it was, then I am afraid it's not going to work. The only possibility for that would be to make some other event (not sure if possible) that will attach such event handler to any new session created during DAG parsing. Not crucial for this PR to get merged, but would be great to see if this is actually ok according to intentions.

Comment (MaksYermak) [reply airflow/dag_processing/processor.py]: @potiuk thank for your investigation! For me your message makes sense. You are right the original idea was to count all queries during DAG file processing. And maybe now I understand why with different event handlers the number of queries are different. I have mentioned it here: https://github.com/apache/airflow/pull/40323#discussion_r1650979649

I have made investigations and recorded queries for event handlers with `session` object and `engine` object. Here's the result:
```
Recorded query locations session:
	processor.py:process_file:883 > processor.py:save_dag_to_db:919 > dagbag.py:_sync_to_db:707 > dagbag.py:_serialize_dag_capturing_errors:672 > serialized_dag.py:write_dag:158:	5
	processor.py:process_file:883 > processor.py:save_dag_to_db:919 > dagbag.py:_sync_to_db:707 > dagbag.py:_serialize_dag_capturing_errors:672 > serialized_dag.py:write_dag:168:	5
	processor.py:process_file:883 > processor.py:save_dag_to_db:919 > dagbag.py:_sync_to_db:707 > dagbag.py:_serialize_dag_capturing_errors:672 > serialized_dag.py:write_dag:181:	5
	dagbag.py:_sync_perm_for_dag:735 > override.py:sync_perm_for_dag:1101 > override.py:create_permission:1836 > override.py:get_permission:1807 > override.py:get_action:1676:	15
	dagbag.py:_sync_perm_for_dag:735 > override.py:sync_perm_for_dag:1101 > override.py:create_permission:1836 > override.py:get_permission:1808 > override.py:get_resource:1737:	15
	dagbag.py:_sync_perm_for_dag:735 > override.py:sync_perm_for_dag:1101 > override.py:create_permission:1839 > override.py:create_resource:1745 > override.py:get_resource:1737:	15
	dagbag.py:_sync_perm_for_dag:735 > override.py:sync_perm_for_dag:1101 > override.py:create_permission:1840 > override.py:create_action:1685 > override.py:get_action:1676:	15
	dagbag.py:_serialize_dag_capturing_errors:679 > dagbag.py:_sync_perm_for_dag:735 > override.py:sync_perm_for_dag:1101 > override.py:create_permission:1836 > override.py:get_permission:1811:	10
	test_processor.py:_process_file:114 > processor.py:process_file:883 > processor.py:save_dag_to_db:919 > dagbag.py:_sync_to_db:710 > dag.py:bulk_write_to_db:3171:	1
	test_processor.py:_process_file:114 > processor.py:process_file:883 > processor.py:save_dag_to_db:919 > dagbag.py:_sync_to_db:710 > dag.py:bulk_write_to_db:3191:	1
	processor.py:process_file:883 > processor.py:save_dag_to_db:919 > dagbag.py:_sync_to_db:710 > dag.py:bulk_write_to_db:3194 > dagrun.py:active_runs_of_dags:380:	1
	processor.py:process_file:883 > processor.py:save_dag_to_db:919 > dagbag.py:_sync_to_db:710 > dag.py:bulk_write_to_db:3263 > dagcode.py:bulk_sync_to_db:82:	1
	python.py:pytest_pyfunc_call:162 > test_processor.py:test_counter_for_last_num_of_db_queries:1022 > test_processor.py:_process_file:114 > processor.py:process_file:893 > processor.py:update_import_errors:618:	2
	python.py:pytest_pyfunc_call:162 > test_processor.py:test_counter_for_last_num_of_db_queries:1022 > test_processor.py:_process_file:114 > processor.py:process_file:893 > processor.py:update_import_errors:625:	1
	test_processor.py:_process_file:114 > processor.py:process_file:904 > processor.py:update_dag_warnings:689 > processor.py:_validate_task_pools:668 > pool.py:get_pools:69:	1
	python.py:pytest_pyfunc_call:162 > test_processor.py:test_counter_for_last_num_of_db_queries:1022 > test_processor.py:_process_file:114 > processor.py:process_file:904 > processor.py:update_dag_warnings:691:	1

Recorded query locations engine:
	processor.py:process_file:883 > processor.py:save_dag_to_db:919 > dagbag.py:_sync_to_db:707 > dagbag.py:_serialize_dag_capturing_errors:672 > serialized_dag.py:write_dag:158:	5
	processor.py:process_file:883 > processor.py:save_dag_to_db:919 > dagbag.py:_sync_to_db:707 > dagbag.py:_serialize_dag_capturing_errors:672 > serialized_dag.py:write_dag:168:	5
	processor.py:process_file:883 > processor.py:save_dag_to_db:919 > dagbag.py:_sync_to_db:707 > dagbag.py:_serialize_dag_capturing_errors:672 > serialized_dag.py:write_dag:181:	5
	dagbag.py:_sync_perm_for_dag:735 > override.py:sync_perm_for_dag:1101 > override.py:create_permission:1836 > override.py:get_permission:1807 > override.py:get_action:1676:	15
	dagbag.py:_sync_perm_for_dag:735 > override.py:sync_perm_for_dag:1101 > override.py:create_permission:1836 > override.py:get_permission:1808 > override.py:get_resource:1737:	15
	dagbag.py:_sync_perm_for_dag:735 > override.py:sync_perm_for_dag:1101 > override.py:create_permission:1839 > override.py:create_resource:1745 > override.py:get_resource:1737:	15
	dagbag.py:_serialize_dag_capturing_errors:679 > dagbag.py:_sync_perm_for_dag:735 > override.py:sync_perm_for_dag:1101 > override.py:create_permission:1839 > override.py:create_resource:1751:	10 -> new
	dagbag.py:_sync_perm_for_dag:735 > override.py:sync_perm_for_dag:1101 > override.py:create_permission:1840 > override.py:create_action:1685 > override.py:get_action:1676:	15
	dagbag.py:_sync_to_db:707 > dagbag.py:_serialize_dag_capturing_errors:679 > dagbag.py:_sync_perm_for_dag:735 > override.py:sync_perm_for_dag:1101 > override.py:create_permission:1845:	15 -> new
	dagbag.py:_serialize_dag_capturing_errors:679 > dagbag.py:_sync_perm_for_dag:735 > override.py:sync_perm_for_dag:1101 > override.py:create_permission:1836 > override.py:get_permission:1811:	10
	test_processor.py:_process_file:114 > processor.py:process_file:883 > processor.py:save_dag_to_db:919 > dagbag.py:_sync_to_db:710 > dag.py:bulk_write_to_db:3171:	1
	test_processor.py:_process_file:114 > processor.py:process_file:883 > processor.py:save_dag_to_db:919 > dagbag.py:_sync_to_db:710 > dag.py:bulk_write_to_db:3191:	1
	processor.py:process_file:883 > processor.py:save_dag_to_db:919 > dagbag.py:_sync_to_db:710 > dag.py:bulk_write_to_db:3194 > dagrun.py:active_runs_of_dags:380:	1
	processor.py:process_file:883 > processor.py:save_dag_to_db:919 > dagbag.py:_sync_to_db:710 > dag.py:bulk_write_to_db:3263 > dagcode.py:bulk_sync_to_db:82:	1
	processor.py:process_file:883 > processor.py:save_dag_to_db:919 > dagbag.py:_sync_to_db:710 > dag.py:bulk_write_to_db:3324 > manager.py:create_datasets:56:	2 -> new
	python.py:pytest_pyfunc_call:162 > test_processor.py:test_counter_for_last_num_of_db_queries:1022 > test_processor.py:_process_file:114 > processor.py:process_file:893 > processor.py:update_import_errors:618:	1
	python.py:pytest_pyfunc_call:162 > test_processor.py:test_counter_for_last_num_of_db_queries:1022 > test_processor.py:_process_file:114 > processor.py:process_file:893 > processor.py:update_import_errors:625:	1
	test_processor.py:_process_file:114 > processor.py:process_file:904 > processor.py:update_dag_warnings:689 > processor.py:_validate_task_pools:668 > pool.py:get_pools:69:	1
	python.py:pytest_pyfunc_call:162 > test_processor.py:test_counter_for_last_num_of_db_queries:1022 > test_processor.py:_process_file:114 > processor.py:process_file:904 > processor.py:update_dag_warnings:691:	1

```
As I see the mostly new queries come from `dagbag`. Maybe as a solution we can change the event handler from `session` to `engine`  and it will make our counter more accurate.

Comment (potiuk) [reply airflow/dag_processing/processor.py]: Yeah. If we can use engine, then yeah, that would be a good one :).

I think you can do it independently from that PR?

Comment (MaksYermak) [reply airflow/dag_processing/processor.py]: @potiuk yes I can do it in another PR

Comment (potiuk) [reply airflow/dag_processing/processor.py]: The context manager approach I introduced looks like a very nice approach for that BTW. I thinl the way I changed it where each DB method counts queries separately (also to account for the fact it can be remote). And there actually - session is perfectly fine). The DAGBag parsing should be additionally wrapped with "engine" event I think to get the queries counted while parsing (but it should be done conditionally - only when database isolation is disabled (because there - there will be no engine to register event in the dag file processor. 

So you will likely need a context manager and wrap parsing with it and that context manager should register event and retrieve the count if db_isolation is off. 

Does it make sense @MaksYermak  ? Would you follow-up with it (I am now working on runinng db isolation tests and we will need that one rather quickly if we want to make it for Airflow 2.10.

Comment (potiuk) [reply airflow/dag_processing/processor.py]: Perfect. I will merge it then and follow up with the rest of AIP-44.

**Comment by MaksYermak** [comment]
@potiuk @kosteev  FYI I have run these changes with `breeze` env for checking the DB queries count function. And after these changes  we have the same number, is 5, for each DAG file, before these changes each DAG file has a different number. I thin these changes somehow broke our `Last # of DB Queries` stats.

**Comment by potiuk** [comment]
> @potiuk @kosteev FYI I have run these changes with breeze env for checking the DB queries count function. And after these changes we have the same number, is 5, for each DAG file, before these changes each DAG file has a different number. I think these changes somehow broke our Last # of DB Queries stats.

Yeah. The whole DAG parsing is now "outside" of the context manager. And I am quite baffled to why the session would be used even if it is not passed down. It should get fixed if we wrap DAG parsing with the session (again conditionally on internal_api) and possibly even more accurate if that wrapping is done with the engine.

**Pull request merged**
